Add UI for user to control build type

DangMinhTam382 opened the pull request:
User can control the value of the CMake built type (CMAKE_BUILD_TYPE), for example Debug or Release via new UI in CMakeBuildTab

Addresses Issue: CDT CMake Improvements #1000, IDE-82683-REQ-013

![New UI for selecting build type](https://github.com/user-attachments/assets/028de1ac-4b19-4bdc-bf45-73c2f8e4ff36)

DangMinhTam382 committed:
Add UI for user to control build type

 User can control the value of the CMake built type (CMAKE_BUILD_TYPE),
for example Debug or Release via new UI in CMakeBuildTab
 Added UT for ICMakeProperties.getBuildType()

Addresses Issue: CDT CMake Improvements #1000, IDE-82683-REQ-013

jonahgraham commented on cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/CMakeBuildTab.java:
```diff
@@ -252,11 +313,29 @@ protected void restoreProperties(Map<String, String> properties) {
+		buildTypeCombo.setText(buildType);
+
+		usedForLaunchModeLabel.setText(MessageFormat.format(Messages.CMakeBuildTab_UsedForLaunchMode,
+				getBuildConfiguration().getLaunchMode()));
```
This should be using the launch mode label, see org.eclipse.debug.core.ILaunchMode.getLabel()

DangMinhTam382 replied:
Updated as suggestion!

jonahgraham commented:
Thanks @DangMinhTam382 I have completed my review.

I have one open question - and that is on the behaviour of the checkbox (related to the spec). I have a meeting scheduled with @betamaxbandit on that tomorrow. I am having trouble putting my question into words right now, but I may have an update request depending on that conversation.

DangMinhTam382 commented on cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/CMakeBuildConfiguration.java:
```diff
@@ -571,7 +574,8 @@ public Map<String, String> getDefaultProperties() {
 				CMAKE_ALL_TARGET, CMAKE_ALL_TARGET_DEFAULT, //
-				CMAKE_CLEAN_TARGET, CMAKE_CLEAN_TARGET_DEFAULT //
+				CMAKE_CLEAN_TARGET, CMAKE_CLEAN_TARGET_DEFAULT, //
+				CMAKE_USE_DEFAULT_BUILD_TYPE, CMAKE_USE_DEFAULT_BUILD_TYPE_DEFAULT //
```
I updated as your suggestion.
Though I have a question:
As I'm understanding, this buildtype is dynamic and is based on LaunchMode, then why should it be init in a place contains static values?

Couldn't we set this value in `CMakeBuildConfiguration.getCMakeProperties()` as done in my commit?
[
(https://github.com/DangMinhTam382/cdt/blob/eeae473811935f4177ceb2215084509e4e0c0aba/cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/CMakeBuildConfiguration.java#L142C3-L149C4)](https://github.com/DangMinhTam382/cdt/blob/9c3a1fe2fc7c985a1c33019dc30032a1ee0c06c5/cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/CMakeBuildConfiguration.java#L142C1-L150C43)

DangMinhTam382 commented on cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/CMakeBuildTab.java:
```diff
@@ -35,6 +37,8 @@
 
 public class CMakeBuildTab extends CommonBuildTab {
 
+	private static final String[] buildTypes = { "Debug", "Release", "RelWithDebInfo", "MinSizeRel", "-custom-" }; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$ //$NON-NLS-5$
```
ah, I saw your reply on my comment about this "custom" value as well.
-
"-custom-" value removed

DangMinhTam382 commented on cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/ICMakeBuildConfiguration.java:
```diff
@@ -70,6 +70,22 @@ public interface ICMakeBuildConfiguration {
+
+	/**
+	 * Name of the build type set by user. Default of this value will be set based on Launch Mode
+	 * (i.e. Debug for Debug launch mode, Release for Run launch mode)
```
Thanks for pointing this out!
Code updated.

DangMinhTam382 commented on cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/messages.properties:
```diff
@@ -7,8 +9,11 @@ CMakeBuildTab_Generator=Generator
 CMakeBuildTab_Settings=CMake Settings
 CMakeBuildTab_Toolchain=Toolchain
+CMakeBuildTab_UseDefaultBuildType=Use default build type
+CMakeBuildTab_UseDefaultBuildType_Tooltip=The default build type is 'debug' for Debug launch mode and 'release' for Run launch mode.
```
Code updated!

DangMinhTam382 commented:
Hi Mr. @betamaxbandit , Mr. @jonahgraham ,

I fixed your comments and added some UT for my changes.
Could you have a look at it again?

Thanks

jld01 commented:
> ![New UI for selecting build type](https://private-user-images.githubusercontent.com/80306525/413953578-180f69e2-5e7a-473d-ad87-d1ea1fb1d1f1.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4NjY5NTIsIm5iZiI6MTczOTg2NjY1MiwicGF0aCI6Ii84MDMwNjUyNS80MTM5NTM1NzgtMTgwZjY5ZTItNWU3YS00NzNkLWFkODctZDFlYTFmYjFkMWYxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE4VDA4MTczMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhMWQyOTI1ZWNhMTY5ZDI0MzE1NGE5ZWQzNDNmNjQ4NDkwZWM0OTYxZTZjOGJlNGE3NGNjYWQ4YTBhYWQwNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bSleBzKk8JLiTfaXjQggvp9Lgerzk4LrIKLh4gEdLdU)

What is the purpose of the _Delete_ button? It seems strange to locate this in the button bar.

DangMinhTam382 commented:
Hi Mr. @jld01 ,

> New UI for selecting build type
What is the purpose of the Delete button? It seems strange to locate this in the button bar.

This seems to be existed button, and not affected by / related to this PR.
Below is where it is created (I think)
https://github.com/eclipse-cdt/cdt/blob/d01f667c21f83a46d7724d05aed768b88c304726/launchbar/org.eclipse.launchbar.ui/src/org/eclipse/launchbar/ui/internal/LaunchBarLaunchConfigDialog.java#L276-L277

TBH, I'm not sure what this button does, when pressed, dialog prompts to delete the launch configuration I assume.
However, when reopen "Edit configuration", previous settings are kept.

Maybe Mr. @jonahgraham or Mr. @betamaxbandit could know something about this one.

jonahgraham commented:
@jld01:

> What is the purpose of the _Delete_ button? It seems strange to locate this in the button bar.

As @DangMinhTam382 mentioned, not directly related to this PR, but the delete button is to match the Delete option in the big Debug Configurations GUI:

![image](https://github.com/user-attachments/assets/d79fa295-7ecc-47b6-a9ea-4d9bd6bf0963)

There is also a Duplicate sometimes for the same reason.

The logic of when to display them is in [LaunchBarLaunchConfigDialog.createConfigButtons(Composite)](https://github.com/eclipse-cdt/cdt/blob/efbba15eabb20843206e6a7c0ec71952e8b10410/launchbar/org.eclipse.launchbar.ui/src/org/eclipse/launchbar/ui/internal/LaunchBarLaunchConfigDialog.java#L244)

This area was most recently updated in https://bugs.eclipse.org/bugs/show_bug.cgi?id=560554 and interestingly that button can show Delete or Restore Defaults. Restore Defaults make be more appropriate verbiage for CMake configs? 

> However, when reopen "Edit configuration", previous settings are kept.

This is sort of correct. The Delete button deletes *only* the launch configuration, not the C build configuration. So pressing Delete (Restore Defaults) restores the defaults for Main, Arguments, and GDB tabs because those are stored in the .launch file. But it doesn't restore defaults for the build settings because they are stored in preferences.

jonahgraham commented:
I have moved the Delete conversation to https://github.com/eclipse-cdt/cdt/issues/1091 and hidden comments on that issue.

jonahgraham reviewed (changes requested):
> As I'm understanding, this buildtype is dynamic and is based on LaunchMode, then why should it be init in a place contains static values?

This is also closely related to the open question I have https://github.com/eclipse-cdt/cdt/pull/1090#issuecomment-2664332363

betamaxbandit commented:
Hi @DangMinhTam382, @jonahgraham,
I just tried this out. Looks really good.

Regarding the question about whether the "Use default build type" checkbox is actually useful, having tried it out, I don't think it is and I think it should be removed. Then it will be consistent with how the Generator combo works (it too supports custom free text values).

betamaxbandit commented on cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/CMakeBuildTab.java:
```diff
@@ -121,6 +130,38 @@ public void widgetSelected(SelectionEvent e) {
+		buildTypeLabel.setText(Messages.CMakeBuildTab_BuildType);
+		buildTypeLabel.setToolTipText(Messages.CMakeBuildTab_BuildType_Tooltip);
+
+		buildTypeCombo = new Combo(buildTypeComp, SWT.DROP_DOWN);
```
How about adding a tooltip to the combo? Something like:

"Typical build type values are shown. A custom type can be entered."

DangMinhTam382 replied:
Tooltip added:
"Typical build types (CMAKE_BUILD_TYPE) are listed. A custom build type can be entered."

jonahgraham commented:
Thanks @betamaxbandit for the review.

@DangMinhTam382 are you able to remove the combo and cooresponding preference `CMAKE_USE_DEFAULT_BUILD_TYPE` by end of Wednesday?

DangMinhTam382 commented on cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/CMakeBuildTab.java:
```diff
@@ -252,11 +314,29 @@ protected void restoreProperties(Map<String, String> properties) {
+		buildTypeCombo.setText(buildType);
+
+		usedForLaunchModeLabel.setText(MessageFormat.format(Messages.CMakeBuildTab_UsedForLaunchMode, DebugPlugin
+				.getDefault().getLaunchManager().getLaunchMode(getBuildConfiguration().getLaunchMode()).getLabel()));
```
Added null check for this code block.
`getBuildConfiguration().getLaunchMode()` will be used in case `getLaunchMode()` return `null`

DangMinhTam382 commented:
Hi Mr. @jonahgraham , Mr. @betamaxbandit ,

> @DangMinhTam382 are you able to remove the **combo** and cooresponding preference CMAKE_USE_DEFAULT_BUILD_TYPE by end of Wednesday?

I assumed "combo" you meant here is the checkBox "Use default build type" and its preference (Hope I assume correctly)

I removed "Use default build type" checkbox and updated the UT.

jonahgraham commented:
> Hi Mr. @jonahgraham , Mr. @betamaxbandit ,
> 
> > @DangMinhTam382 are you able to remove the **combo** and cooresponding preference CMAKE_USE_DEFAULT_BUILD_TYPE by end of Wednesday?
> 
> I assumed "combo" you meant here is the checkBox "Use default build type" and its preference (Hope I assume correctly)

Yes.! Sorry for the incorrect comment


> 
> I removed "Use default build type" checkbox and updated the UT.

Great! Thank you 

>

jonahgraham reviewed (approved):
Thanks @DangMinhTam382 - once CI passes this is good to go

betamaxbandit reviewed (approved):
LGTM
I especially like the JUnits in [CMakeBuildConfigurationTests.java](https://github.com/eclipse-cdt/cdt/pull/1090/files#diff-a5e4b39d1cd8007a06c0a7b63a397ae832963944c8572074166691aa05c6d069)

jonahgraham commented:
Thanks @DangMinhTam382 

I updated the screenshot in the original description to reflect what where we ended up.
